### DISCUSSION TITLE
Help with Integrating LLVM

### DISCUSSION THREAD

**Discussion by jiezhuzzz** [cpm-cmake/CPM.cmake#620] [Q&A] [answered]
Hi everyone, I have trouble integrating the LLVM package into my project. Here’s the segment I used in my CMakeLists.txt:
```
CPMAddPackage(
  NAME llvm
  VERSION 18.1.8
  URL https://github.com/llvm/llvm-project/releases/download/llvmorg-18.1.8/llvm-project-18.1.8.src.tar.xz
)
```
When I wrote the following code,

```c++
#include <llvm/IR/PassManager.h>
namespace llvm {

class HelloWorldPass : public PassInfoMixin<HelloWorldPass> {
public:
  PreservedAnalyses run(Function &F, FunctionAnalysisManager &AM);
};
} // namespace llvm
```
I got the following build output:
```
[build] <some_path_to_cpp_file>:2:10: fatal error: 'llvm/IR/PassManager.h' file not found
[build]     2 | #include <llvm/IR/PassManager.h>
[build]       |          ^~~~~~~~~~~~~~~~~~~~~~~
[build] 1 error generated.
```
Has anyone integrated LLVM into their project before? How did you set the configuration?

**Comment by Avus-c** [comment]
One immediate issue is that the llvm-project repository does not have a top-level CMakeLists.txt file. To resolve this, you need to use the SOURCE_SUBDIR argument to specify the location of the CMakeLists.txt file within the repository.

Since I am not deeply familiar with the llvm-project structure, there might be additional adjustments required.

**Reply by jiezhuzzz** [reply]
Hi Avus-c, thanks for your reply! I added the `SOURCE_SUBDIR` options, and now my project can automatically compile LLVM. However, I still cannot use LLVM APIs in my project. I checked the official LLVM [documentation](https://llvm.org/docs/CMake.html#embedding-llvm-in-your-project) on how to embed LLVM with CMake, but the procedure seems different from the examples in the wiki.

I want to explore and add examples of installing LLVM to the wikis. However, I'm new to C++ and CMake. Would you mind giving me some hints on which part of the demo CMakeList I should refer to? Thanks!

**Reply by Avus-c** [reply, marked as answer]
Sorry for the late response.

I'm not very familiar with LLVM's setup, and unfortunately, I don't have the time to look into it in detail. Sorry about that.

However, my intuition is that LLVM may not be easily usable with CPM, at least without significant configuration effort. From what I’ve seen in the examples, LLVM seems to be designed for integration using `find_package`. The recommended approach would be to install LLVM separately so that `find_package` can locate the installed version. After that, you can directly use the binaries, libraries, ... in your project.

CPM, on the other hand, is great for adding dependencies that you can built alongside your project. This approach works well for small projects and simple CMake configurations but can become a disadvantage for larger projects with complex build configrations like LLVM. In my experience, a CMake project usually needs to be designed for source-level integration for this to work seamlessly.

That said, don't be discouraged to try. :D
It might be easier than I anticipate.
